dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala

```diff
@@ -429,17 +452,34 @@ object DistriOptimizer {
       val workingModels = modelIter.next().localModels
       workingModels.foreach(_.evaluate())
       dataIter.map(batch => {
-        require(batch.data.size(1) == batch.labels.size(1))
-        val stackSize = batch.data.size(1) / _subModelNumber
-        val extraSize = batch.data.size(1) % _subModelNumber
+        val tensorBatchType = batch.data match {
+          case tensor: Tensor[T] =>
+            require(batch.data.toTensor[T].size(1) == batch.labels.toTensor[T].size(1))
+            true
+          case table: Table =>
+            require(batch.data.toTable.length == batch.labels.toTable.length)
+            false
+        }
+        val (stackSize, extraSize) = tensorBatchType match {
+          case true => (batch.data.toTensor[T].size(1) / _subModelNumber,
+            batch.data.toTensor[T].size(1) % _subModelNumber)
+          case false => (batch.data.toTable.length / _subModelNumber,
+            batch.data.toTable.length % _subModelNumber)
+        }
         val parallelism = if (stackSize == 0) extraSize else _subModelNumber
         Engine.default.invokeAndWait(
           (0 until parallelism).map(b =>
             () => {
               val offset = b * stackSize + math.min(b, extraSize)
               val length = stackSize + (if (b < extraSize) 1 else 0)
-              val input = batch.data.narrow(1, offset + 1, length)
-              val target = batch.labels.narrow(1, offset + 1, length)
+              val input = tensorBatchType match {
+                case true => batch.data.toTensor[T].narrow(1, offset + 1, length)
+                case false => NarrowTable[T](offset + 1, length).updateOutput(batch.data.toTable)
+              }
+              val target = tensorBatchType match {
+                case true => batch.labels.toTensor[T].narrow(1, offset + 1, length)
+                case false => NarrowTable(offset + 1, length).updateOutput(batch.labels.toTable)
+              }
```

Comment:
It can also be defined as something like:
val (input, target) = tensorBatchType match {
                case true => (batch.data.toTensor.narrow(1, offset + 1, length),
                                       batch.labels.toTensor.narrow(1, offset + 1, length)
                case false => (narrowTable.updateOutput(batch.data.toTable),
                                       narrowTable.updateOutput(batch.labels.toTable))
              }